gtsam/slam/SmartProjectionFactorP.h

```diff
@@ -0,0 +1,359 @@
+class SmartProjectionFactorP : public SmartProjectionFactor<CAMERA> {
+
+ private:
+  typedef SmartProjectionFactor<CAMERA> Base;
+  typedef SmartProjectionFactorP<CAMERA> This;
+  typedef typename CAMERA::CalibrationType CALIBRATION;
+
+  static const int DimPose = 6;  ///< Pose3 dimension
+  static const int ZDim = 2;  ///< Measurement dimension (Point2)
+
+ protected:
+
+  /// vector of keys (one for each observation) with potentially repeated keys
+  std::vector<Key> nonUniqueKeys_;
+
+  /// shared pointer to calibration object (one for each observation)
+  std::vector<boost::shared_ptr<CALIBRATION> > K_all_;
+
+  /// Pose of the camera in the body frame (one for each observation)
+  std::vector<Pose3> body_P_sensors_;
+
+ public:
+  typedef CAMERA Camera;
+  typedef CameraSet<CAMERA> Cameras;
+  typedef Eigen::Matrix<double, ZDim, DimPose> MatrixZD;  // F blocks
+  typedef std::vector<MatrixZD, Eigen::aligned_allocator<MatrixZD> > FBlocks;  // vector of F blocks
+
+  /// shorthand for a smart pointer to a factor
+  typedef boost::shared_ptr<This> shared_ptr;
+
+  /// Default constructor, only for serialization
+  SmartProjectionFactorP() {
+  }
+
+  /**
+   * Constructor
+   * @param sharedNoiseModel isotropic noise model for the 2D feature measurements
+   * @param params parameters for the smart projection factors
+   */
+  SmartProjectionFactorP(const SharedNoiseModel& sharedNoiseModel,
+                         const SmartProjectionParams& params =
+                             SmartProjectionParams())
+      : Base(sharedNoiseModel, params) {
+    // use only configuration that works with this factor
+    Base::params_.degeneracyMode = gtsam::ZERO_ON_DEGENERACY;
+    Base::params_.linearizationMode = gtsam::HESSIAN;
+  }
+
+  /** Virtual destructor */
+  ~SmartProjectionFactorP() override {
+  }
+
+  /**
+   * add a new measurement, corresponding to an observation from pose "poseKey" whose camera
+   * has intrinsic calibration K and extrinsic calibration body_P_sensor.
+   * @param measured 2-dimensional location of the projection of a
+   * single landmark in a single view (the measurement)
+   * @param poseKey key corresponding to the body pose of the camera taking the measurement
+   * @param K (fixed) camera intrinsic calibration
+   * @param body_P_sensor (fixed) camera extrinsic calibration
+   */
+  void add(const Point2& measured, const Key& poseKey,
+           const boost::shared_ptr<CALIBRATION>& K, const Pose3 body_P_sensor =
+               Pose3::identity()) {
+    // store measurement and key
+    this->measured_.push_back(measured);
+    this->nonUniqueKeys_.push_back(poseKey);
+
+    //  also store keys in the keys_ vector: these keys are assumed to be unique, so we avoid duplicates here
+    if (std::find(this->keys_.begin(), this->keys_.end(), poseKey) == this->keys_.end())
+      this->keys_.push_back(poseKey);  // add only unique keys
+
+    // store fixed intrinsic calibration
+    K_all_.push_back(K);
```

Comment:
I am a bit worried about the design. Think about 10,000 time steps of a 6 camera rig. Will we have 60,000 calibrations? And 60,000 body_P_sensors?

Reply:
solved as discussed